### PR TITLE
Parallel-CF

### PR DESCRIPTION
**Pull request by jan-smartward**
Allow async CF deployment and then waiting as a seperate call.

### CODE REVIEW

File: vars/cloudformation.groovy [hunk 1]
```diff
@@ -75,58 +76,79 @@ def call(body) {
...
-      success = true // Non-successes would have already thrown an exception in recursive calls
+def handleActionRequest(cf, config){
+  def assertOrThrow = { booleanAssertion ->
+    if (!booleanAssertion) {
+      printFailedStackEvents(cf, config.stackName, config.region)
+      throw new Exception("Stack ${config.stackName} failed to ${config.action}")
+    }
+  }
+  
+  // Should start an action
```
Comment (sw-phil) [inline vars/cloudformation.groovy:87]: There is a confusing mixture of:
- setting the `env` variables
- printing status
- returning

Depending on action and states. Seems messy and it is hard to figure out if these differences are intentional and/or important.

File: vars/cloudformation.groovy [hunk 2]
```diff
@@ -370,10 +392,18 @@ def wait(cf, stackName, successStatus)   {
...
-def doesStackExist(cf, stackName) {
+def doesStackExist(cf, stackName, inState = null) {
   try {
     DescribeStacksResult result = cf.describeStacks(new DescribeStacksRequest().withStackName(stackName))
-    return result != null
+    if (result != null) {
+      if (inState == null) {
+        return true
+      } else {
+        return result.getStacks().get(0).getStackStatus() == inState
```
Comment (sw-phil) [inline vars/cloudformation.groovy:402]: If getStacks().len != 1: Should this be reported / logged?

**Review by sw-phil — APPROVED** [review]

File: vars/cloudformation.groovy (hunk 2 above)
Comment (jan-smartward) [reply vars/cloudformation.groovy:402]: That will presumably return an exception when `.get(0)` fails.

File: vars/cloudformation.groovy (hunk 1 above)
Comment (jan-smartward) [reply vars/cloudformation.groovy:87]: Yes, it kind of started that way though. If we'd want to clean that up, we'd probably have to redo this a bit and then not have it be compatible with the upstream anymore?

**Pull request merged**